test/base.py

```diff
@@ -34,194 +35,242 @@
+    """Create the base test case class for package tests
+
+    This function produces the base class for qiskit-experiments tests using
+    either ``unittest.TestCase`` or ``testtools.TestCase`` for the base class.
+    The creation of the class is done in this function rather than directly
+    executed in the module so that, even when ``USE_TESTTOOLS`` is true, a
+    ``unittest`` base class can be produced for ``test_base.py`` to check that
+    no hard-dependence on ``testtools`` has been introduced.
+    """
+    if use_testtools:
+
+        class BaseTestCase(testtools.TestCase):
+            """Base test class."""
+
+            # testtools maintains their own version of assert functions which mostly
+            # behave as value adds to the std unittest assertion methods. However,
+            # for assertEquals and assertRaises modern unittest has diverged from
+            # the forks in testtools and offer more (or different) options that are
+            # incompatible testtools versions. Just use the stdlib versions so that
+            # our tests work as expected.
+            assertRaises = unittest.TestCase.assertRaises
+            assertEqual = unittest.TestCase.assertEqual
+
+    else:
+
+        class BaseTestCase(unittest.TestCase):
+            """Base test class."""
+
+            def useFixture(self, fixture):  # pylint: disable=invalid-name
+                """Shim so that useFixture can be called in subclasses
+
+                useFixture is a testtools.TestCase method. The actual fixture is
+                not used when using unittest.
+                """
+
+    class QETestCase(BaseTestCase):
+        """Qiskit Experiments specific extra functionality for test cases."""
+
+        def setUp(self):
+            super().setUp()
+            self.useFixture(fixtures.Timeout(TEST_TIMEOUT, gentle=True))
+
+        @classmethod
+        def setUpClass(cls):
+            """Set-up test class."""
+            super().setUpClass()
+
+            warnings.filterwarnings("error", category=DeprecationWarning)
+
+            # Some functionality may be deprecated in Qiskit Experiments. If
+            # the deprecation warnings aren't filtered, the tests will fail as
+            # ``QiskitTestCase`` sets all warnings to be treated as an error by
+            # default.
+            # pylint: disable=invalid-name
+            allow_deprecationwarning_message = [
+                # TODO: Remove in 0.6, when submodule `.curve_analysis.visualization` is removed.
+                r".*Plotting and drawing functionality has been moved",
+                r".*Legacy drawers from `.curve_analysis.visualization are deprecated",
+            ]
+            for msg in allow_deprecationwarning_message:
+                warnings.filterwarnings("default", category=DeprecationWarning, message=msg)
+
+        def assertExperimentDone(
+            self,
+            experiment_data: ExperimentData,
+            timeout: float = 120,
```

Comment:
Maybe this could be tied to `TEST_TIMEOUT`? At least it doesn't make sense to be 120 if `TEST_TIMEOUT` is 60.

Reply:
Hmm, tying this to `TEST_TIMEOUT` seems kind of reasonable. In practice that will never be reached unless there is some timing uncertainty in the `signal.alarm` callback, but it would work for the case when not using testtools. Another option might be to get rid of this timeout and just rely on testtools time out. I think this may have been an earlier attempt to do what the time out fixture is doing now. What do you think?

Also, this time out is not really a change in this PR -- this is existing code that just got shifted down in scope.

Reply:
Yeah, I only noticed it when reviewing this PR (and I also just noticed a few cases in tests where a large timeout value is being manually passed, which is also effectively not doing anything). I think the parameter could be worth keeping in case someone wants a timeout while not using testtools, and setting it to `TEST_TIMEOUT` may be a bit easier to use. I'm not very hung up over it either way.